README
MojoX-Renderer-TT

This module implements a Template Toolkit renderer for the Mojo
framework.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc MojoX::Renderer::TT

You can also look for information at:

    Git repository
        http://git.develooper.com/?p=MojoX-Renderer-TT.git;a=summary
        git://git.develooper.com/MojoX-Renderer-TT.git

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=MojoX-Renderer-TT

    CPAN Ratings
        http://cpanratings.perl.org/d/MojoX-Renderer-TT

    Search CPAN
        http://search.cpan.org/dist/MojoX-Renderer-TT/


COPYRIGHT AND LICENCE

Copyright (C) 2008-2009 Ask Bjørn Hansen

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



Changes
Revision history for Mojolicious::Plugin::TtRenderer

{{$NEXT}}

1.62      2021-03-15 04:32:27 -0600
  - Documentation improvements (gh#53)
  - Moved to new GitHub org

1.61      2020-09-02 06:05:42 -0600
  - Require Perl 5.16
  - Free stash when request is done, instead of during the next request (gh#51, mattlaw++)

1.60      2018-10-16 09:57:46 -0400
  - Template 2.28 compatability

1.59      2017-07-18 21:57:51 -0400
  - Mojolicious 7.31 compatability

1.58      2017-03-14 11:01:17 -0400
  - More mojolicious deprecations

1.57      2015-11-23 17:16:39 -0500
  - Mojolicious 6.33 compatability
  - Require Mojolicious 6.33

1.56      2015-04-13 12:29:00 -0400
  - Only force template compiling if not specified by caller (GH#43 dlage++)

1.55      2015-02-28 10:31:58 -0500
  - Mojo 6.x compat  

1.54      2014-12-20 10:17:47 -0500
  - cpantesters: do not report on older perls

1.53      2014-05-31 06:22:47 -0400
  - Mojolicious 5.0 support

1.52      2014-05-08 19:26:47 -0400
  - updated repository name

1.51      2014-03-21 14:59:39 -0400
  - extended diagnostics
  - documentation tweaks
  - fix non-lite app example (GH#41, uwisser)

1.50      2013-08-08T16:51:21-0400
  - update Changes to follow spec (GH#39, Сергей Романов)
  - minor documentation fixes

1.49    2013-08-08T09:18:54-0400
  - documentation

1.48    2013-08-08T09:13:19-0400
  - documentation

1.47    2013-07-26T06:07:35-0400
  - Template-Toolkit 2.25 compatability

1.46    2013-05-19
  - Mojolicious 4.0 compatability

1.45    2013-04-26
  - correctly support app->default->{layout} (GH#37, Fayland Lam)

1.44    2013-04-13
  - promote development changes to production

1.43_01 2013-04-11
  - _template_modified returns value from super class instead of 1 if true (GH#36, Fayland Lam)

1.43    2013-04-08
  - promote development changes

1.42_04 2013-04-06
  - Updated meta for new github repo location

1.42_03 2013-03-22
  - Allow custom relative paths for cache_dir (GH#35, Jason Crowther)

1.42_02 2013-02-26
  - only create COMPILE_DIR based on UID if the user doesn't specify COMPILE_DIR

1.42_01 2013-02-24
  - set COMPILE_DIR based on UID so that multiple users can use TtRenderer with default config

1.42    2013-02-24
  - fix for test on solaris

1.41    2013-02-23
  - Additional testing diagnostics

1.40    2013-02-20
  - Added diagnostic for IO loop implementation in testing

1.39    2013-01-28
  - Documentation updates.

1.38    2013-01-27
  - fixed but where DATA templates were not being rendered with empty renderer path

1.37    2013-01-12
  - fix default_template2.t to use temp directory for COMPILE_DIR

1.36    2013-01-09
  - Use $provider->fetch to find templates (GH#34)

1.35    2012-12-30
  - Documentation fix

1.34    2012-12-29
  - Include META.json in distribution.

1.33    2012-12-27
  - Fix hang in t/deep_recursion.t on *BSD

1.32    2012-12-27
  - silenced a few annoying warnings during test
  - use temp directory for COMPILE_DIR in tests to avoid failures if the
          default compile directory already exists and is owned by someone else.

1.31    2012-12-26
  - Set locale "C" in tests that rely on it

1.30    2012-12-22
  - Don't rely on English locale in the test t/tt_plugin_lite_app.t

1.29    2012-12-18
  - support multiple renderer paths

1.28    2012-10-11
  - specify minimum perl version

1.27    2012-10-02
  - documentation fixes and updates.

1.26    2012-08-31
  - Windows compatability with Cwd::abs_path

1.25    2012-08-25
  - Return 0 on not found, die on error (GH#30)

1.24    2012-08-23
  - Compatibility with Mojolicious 3.33+

1.23    2012-08-23
  - Use die instead of render_exception (GH#29)

1.22    2012-08-21
  - Compatibility with Mojolicious 3.05+ (GH#24)
  - Avoid deep recursion when exception template dies (GH#25,GH26)
  - Fixed test failures (GH#27,GH#28)

1.21    2012-06-05
  - Fixed double TT rendering on error (Matthias Bethke (GH#21))
  - Cache templates in tmpdir by default (Marcus Ramberg)
  - Require Mojolicious 2.51 to avoid memory leaks (GH#19)
  - Document the 'c' stash variable (RT#70866, Matthias Bethke)
  - Fix memory leaks, add tests for them (GH#18, Magnus Holm)
  - Use app->renderer->paths to find template directories (GH#17)

1.20    2011-08-01
  - Drop use of MojoX namespace for Mojolicious::Plugin::TtRenderer::
          (GH#12, Cosimo Streppone)

1.13    2011-06-07
  - Compatibility with Mojolicious 1.3+

1.12    2011-02-28
  - Inline wrappers and includes now work on Windows
          (GH#13, Christiaan Kras)

1.11    2011-02-06
  - Change to use relative paths (Marcus Ramberg)
  - Add exception template for tests

1.10    2011-02-05
  - Support Mojolicious 'layout'/'extends' (Maxim Vuets, Marcus
          Ramberg)
  - Fix Strawberry Perl tests (RT#65282, Christiaan Kras, Ask)
  - Update bugtracker URL in META.yml

1.0     2011-01-22
  - Require (and be compatible with) Mojolicious 1.0+
  - Added inline template support (Viacheslav Tykhanovskyi & Ask)
  - Fix inline rendering and test failures (Marcus Ramberg)
  - Fixed helper compatibility with Mojolicious 1.0 (Árpád Szász)
  - Support new mojolicious error template (Marcus)
  - Fix unicode test (Marcus & Spleenjack)
  - Replaced deprecated Mojo dependence with Mojolicious (spleenjack)

0.40    2010-08-02
  - Updated for newer Mojo (Viacheslav Tykhanovskyi)
  - Added helpers (Viacheslav Tykhanovskyi)
  - Documentation updates

0.31    2009-09-11
  - Remove deprecated 'new' method
  - Update MANIFEST (oops, thanks Sebastian)

0.30    2009-09-09
  - Revert the change of template_path stash key
  - Reflect recent changes in mojo project (v0.991251) (Sebastian Knapp)
  - One test with Mojolicious added (Sebastian Knapp)

0.21    2009-07-29
  - Update stash key from 'template_path' to 'template' to
          match mojo change (Sebastian Knapp)

0.20    2008-12-09
  - Updates to work with Mojo 0.9
  - Add default INCLUDE_PATH
  - Made default compile dir not Unix specific (Viacheslav
          Tykhanovskyi)


t/00_diag.pl
use strict;
use warnings;
use Test::More;
use Mojo::IOLoop;

our $format;

my $loop = eval { Mojo::IOLoop->singleton };
diag $@ if $@;
diag sprintf $format, 'mojo io loop', ref eval { $loop->reactor };
diag $@ if $@;

1;


t/templates/Foo.pm
package
  Foo;
use strict;
use warnings;
1;
__DATA__

@@ include.inc
Hello
@@ includes/sub/include.inc
Hallo

@@ layouts/layout.html.tt
w[%- content -%]d



lib/MojoX/Renderer/TT.pm
package MojoX::Renderer::TT;

use warnings;
use strict;

use base 'Mojo::Base';

use File::Spec ();
use Mojo::ByteStream 'b';
use Template ();
use Cwd qw/abs_path/;

__PACKAGE__->attr('tt');

sub build {
    my $self = shift->SUPER::new(@_);
    $self->_init(@_);
    return sub { $self->_render(@_) }
}

sub _init {
    my $self = shift;
    my %args = @_;

    #$Template::Parser::DEBUG = 1;

    my $app = delete $args{mojo} || delete $args{app};

    my $dir = $app && $app->home->rel_dir('tmp/ctpl');

    # TODO
    #   take and process options :-)

    my %config = (
        ($app ? (INCLUDE_PATH => abs_path($app->home->rel_dir('templates'))) : ()),
        COMPILE_EXT => '.ttc',
        COMPILE_DIR => ($dir || abs_path(File::Spec->tmpdir)),
        UNICODE     => 1,
        ENCODING    => 'utf-8',
        CACHE_SIZE  => 128,
        RELATIVE    => 1,
        %{$args{template_options} || {}},
    );

    $config{LOAD_TEMPLATES} =
      [MojoX::Renderer::TT::Provider->new(%config, renderer => $app->renderer)]
      unless $config{LOAD_TEMPLATES};

    $self->tt(Template->new(\%config))
      or Carp::croak "Could not initialize Template object: $Template::ERROR";

    return $self;
}

sub _render {
    my ($self, $renderer, $c, $output, $options) = @_;

    # Inline
    my $inline = $options->{inline};

    # Template
    my $t = $renderer->template_name($options);
    $t = 'inline' if defined $inline;
    return unless $t;


    my $helper = MojoX::Renderer::TT::Helper->new(ctx => $c);

    # Purge previous result
    $$output = '';

    my @params = ({%{$c->stash}, c => $c, h => $helper}, $output, {binmode => ':utf8'});
    $self->tt->{SERVICE}->{CONTEXT}->{LOAD_TEMPLATES}->[0]->ctx($c);

    my $ok = $self->tt->process(defined $inline ? \$inline : $t, @params);

    # Error
    unless ($ok) {

        my $e = Mojo::Exception->new(
            $self->tt->error.'',
            $self->tt->service->process(defined $inline ? \$inline : $t));
        $$output = '';
        $c->app->log->error(qq/Template error in "$t": $e/);
        $c->render_exception($e);
        $self->tt->error('');
        return 0;
    }

    return 1;
}

1;    # End of MojoX::Renderer::TT

package
  MojoX::Renderer::TT::Helper;

use strict;
use warnings;

use base 'Mojo::Base';

our $AUTOLOAD;

__PACKAGE__->attr('ctx');

sub AUTOLOAD {
    my $self = shift;

    my $method = $AUTOLOAD;

    return if $method =~ /^[A-Z]+?$/;
    return if $method =~ /^_/;
    return if $method =~ /(?:\:*?)DESTROY$/;

    $method = (split '::' => $method)[-1];

    die qq/Unknown helper: $method/ unless $self->ctx->app->renderer->helpers->{$method};

    return $self->ctx->$method(@_);
}

1;

package
  MojoX::Renderer::TT::Provider;

use strict;
use warnings;

use base 'Template::Provider';

sub new {
    my $class = shift;
    my %params = @_;

    my $renderer = delete $params{renderer};

    my $self = $class->SUPER::new(%params);
    $self->renderer($renderer);
    return $self;
}

sub renderer      { @_ > 1 ? $_[0]->{renderer}      = $_[1] : $_[0]->{renderer} }
sub ctx           { @_ > 1 ? $_[0]->{ctx}           = $_[1] : $_[0]->{ctx} }

sub _template_modified {1}

sub _template_content {
    my $self = shift;
    my ($path) = @_;

    # Convert backslashes to forward slashes to make inline templates work on Windows
    $path =~ s/\\/\//g;
    my ($t) = ($path =~ m{templates\/(.*)$});

    if (-r $path) {
        return $self->SUPER::_template_content(@_);
    }

    # Try DATA section
    elsif (my $d = $self->renderer->get_data_template($self->ctx, $t)) {
        return wantarray ? ($d, '', time) : $d;
    }

    my $data = '';
    my $error = "$path: not found";
    my $mod_date = time;
    return wantarray ? ($data, $error, $mod_date) : $data;
}

1;

__END__

=encoding utf-8

=head1 NAME

MojoX::Renderer::TT - Template Toolkit renderer for Mojo

=head1 SYNOPSIS

Add the handler:

    sub startup {
        ...

        # Via mojolicious plugin
        $self->plugin(tt_renderer => {template_options => {FILTERS => [ ... ]}});

        # Or manually
        use MojoX::Renderer::TT;

        my $tt = MojoX::Renderer::TT->build(
            mojo => $self,
            template_options => {
                PROCESS  => 'tpl/wrapper',
                FILTERS  => [ ... ],
                UNICODE  => 1,
                ENCODING => 'UTF-8',
            }
        );

        $self->renderer->add_handler( tt => $tt );
    }

Template parameter are taken from C< $c->stash >.

=head1 RENDERING

The template file for C<"example/welcome"> would be C<"templates/welcome.html.tt">.

When template file is not available rendering from C<__DATA__> is attempted.

    __DATA__

    @@ welcome.html.tt
    Welcome, [% user.name %]!

Inline template is also supported:

    $self->render(inline => '[% 1 + 1 %]', handler => 'tt');

=head1 HELPERS

Helpers are exported automatically under C<h> namespace.

    [% h.url_for('index') %]

=head1 METHODS

=head2 build

This method returns a handler for the Mojolicious renderer.

Supported parameters are

=over 4

=item mojo
C<build> currently uses a C<mojo> parameter pointing to the base class (Mojo).
object. When used the INCLUDE_PATH will be set to 

=item template_options

A hash reference of options that are passed to Template->new().

=back

=head1 AUTHOR

Ask Bjørn Hansen, C<< <ask at develooper.com> >>

=head1 TODO

   * Better support non-Mojolicious frameworks
   * Move the default template cache directory?
   * Better way to pass parameters to the templates? (stash)
   * More sophisticated default search path?

=head1 BUGS

Please report any bugs or feature requests to C<bug-mojox-renderer-tt at rt.cpan.org>,
or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=MojoX-Renderer-TT>.  I will be
notified, and then you'll automatically be notified of progress on your bug as I
make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc MojoX::Renderer::TT

You can also look for information at:

=over 4

=item * git repository

L<http://git.develooper.com/?p=MojoX-Renderer-TT.git;a=summary>,
L<git://git.develooper.com/MojoX-Renderer-TT.git>

L<http://github.com/abh/mojox-renderer-tt/>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=MojoX-Renderer-TT>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/MojoX-Renderer-TT>

=item * Search CPAN

L<http://search.cpan.org/dist/MojoX-Renderer-TT/>

=back

=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2008-2010 Ask Bjørn Hansen, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut


lib/Mojolicious/Plugin/TtRenderer.pm
package Mojolicious::Plugin::TtRenderer;

use strict;
use warnings;
use 5.016;

# ABSTRACT: Template Renderer Plugin for Mojolicious
# VERSION

use base 'Mojolicious::Plugin';

use Mojolicious::Plugin::TtRenderer::Engine;

sub register {
    my ($self, $app, $args) = @_;

    $args ||= {};

    my $tt = Mojolicious::Plugin::TtRenderer::Engine->build(%$args, app => $app);

    # Add "tt" handler
    $app->renderer->add_handler(tt => $tt);
}

$Mojolicious::Plugin::TtRenderer::VERSION //= ('devel');

1;
__END__

=encoding utf-8

=begin stopwords

Bjørn
Szász
Árpád
Романов
Сергей

=end stopwords

=head1 SYNOPSIS

L<Mojolicious::Lite>:

 plugin 'tt_renderer';

L<Mojolicious>:

 $self->plugin('tt_renderer');

=head1 DESCRIPTION

This plugin is a simple Template Toolkit renderer for L<Mojolicious>.
Its defaults are usually reasonable, although for finer grain detail in
configuration you may want to use
L<Mojolicious::Plugin::TtRenderer::Engine> directly.

=head1 OPTIONS

These are the options that can be passed in as arguments to this plugin.

=head2 template_options

Configuration hash passed into L<Template>'s constructor, see
L<Template Toolkit's configuration summary|Template#CONFIGURATION-SUMMARY>
for details.  Here is an example using the L<Mojolicious::Lite> form:

 plugin 'tt_renderer' => {
   template_options => {
     PRE_CHOMP => 1,
     POST_CHOMP => 1,
     TRIM => 1,
   },
 };

Here is the same example using the full L<Mojolicious> app form:

 package MyApp;
 
 use Mojo::Base qw( Mojolicious );
 
 sub startup {
   my($self) = @_;
 
   $self->plugin('tt_renderer' => {
     template_options => {
       PRE_CHOMP => 1,
       POST_CHOMP => 1,
       TRIM => 1,
     },
   }
 
   ...
 }

These options will be used if you do not override them:

=over 4

=item INCLUDE_PATH

Generated based on your application's renderer's configuration.  It
will include all renderer paths, in addition to search files located
in the C<__DATA__> section by the usual logic used by L<Mojolicious>.

=item COMPILE_EXT

C<.ttc>

=item UNICODE

C<1> (true)

=item ENCODING

C<utf-87>

=item CACHE_SIZE

C<128>

=item RELATIVE

C<1> (true)

=back

=head2 cache_dir

Specifies the directory in which compiled template files are
written.  This:

 plugin 'tt_renderer', { cache_dir => 'some/path' };

is equivalent to

 plugin 'tt_renderer', { template_options { COMPILE_DIR => 'some/path' } };

except in the first example relative paths are relative to the L<Mojolicious>
app's home directory (C<$app-E<gt>home>).

=head1 STASH

=head2 h

Helpers are available via the C<h> entry in the stash.

 <a href="[% h.url_for('index') %]">go back to index</a>

=head2 c

The current controller instance can be accessed as C<c>.

 I see you are requesting a document from [% c.req.headers.host %].

=head1 EXAMPLES

=over 4

=item L<Mojolicious::Lite> example:

# EXAMPLE: example/myapp.pl

=item L<Mojolicious> example:

C<lib/MyApp.pm>:

# EXAMPLE: example/myapp/lib/MyApp.pm

C<lib/MyApp/Example.pm>:

# EXAMPLE: example/myapp/lib/MyApp/Example.pm

C<templates/example/welcome.html.tt>:

# EXAMPLE: example/myapp/templates/example/welcome.html.tt

C<templates/layouts/default.html.tt>:

# EXAMPLE: example/myapp/templates/layouts/default.html.tt

=back

These are also included with the C<Mojolicious::Plugin::TtRenderer>
distribution, including the support files required for the full
L<Mojolicious> app example.

=head1 SEE ALSO

L<Mojolicious::Plugin::TtRenderer::Engine>,
L<Mojolicious>,
L<Mojolicious::Guides>,
L<http://mojolicious.org>.

=cut


lib/Mojolicious/Plugin/TtRenderer/Engine.pm
package Mojolicious::Plugin::TtRenderer::Engine;

use warnings;
use strict;
use 5.016;
use base 'Mojo::Base';
use Carp ();
use File::Spec ();
use Mojo::ByteStream 'b';
use Template ();
use Cwd qw/abs_path/;
use Scalar::Util 'weaken';
use POSIX ':errno_h';

# ABSTRACT: Template Toolkit renderer for Mojolicious
# VERSION

__PACKAGE__->attr('tt');

sub build {
    my $self = shift->SUPER::new(@_);
    weaken($self->{app});
    $self->_init(@_);
    sub { $self->_render(@_) }
}

sub _init {
    my $self = shift;
    my %args = @_;

    #$Template::Parser::DEBUG = 1;

    my $dir;
    my $app = delete $args{mojo} || delete $args{app};
    if($dir=$args{cache_dir}) {

      if($app && substr($dir,0,1) ne '/') {
        $dir=$app->home->rel_file($dir);
      }
    }

    # TODO
    #   take and process options :-)

    my @renderer_paths = $app ? map { abs_path($_) } grep { -d $_ } @{ $app->renderer->paths } : ();
    push @renderer_paths, 'templates';

    my %config = (
        INCLUDE_PATH => \@renderer_paths,
        COMPILE_EXT  => '.ttc',
        UNICODE      => 1,
        ENCODING     => 'utf-8',
        CACHE_SIZE   => 128,
        RELATIVE     => 1,
        %{$args{template_options} || {}},
    );

    if ( !exists( $config{COMPILE_DIR} ) ) {
        $config{COMPILE_DIR} //= $dir || do {
          my $tmpdir = File::Spec->catdir(File::Spec->tmpdir, "ttr$<");
          mkdir $tmpdir unless -d $tmpdir;
          $tmpdir;
        };
    }

    $config{LOAD_TEMPLATES} =
      [Mojolicious::Plugin::TtRenderer::Provider->new(%config, renderer => $app->renderer)]
      unless $config{LOAD_TEMPLATES};

    $self->tt(Template->new(\%config))
      or Carp::croak "Could not initialize Template object: $Template::ERROR";

    $self;
}

sub _render {
    my ($self, $renderer, $c, $output, $options) = @_;

    # Inline
    my $inline = $options->{inline};

    # Template
    my $t = $renderer->template_name($options);
    $t = 'inline' if defined $inline;
    return unless $t;

    my $helper = Mojolicious::Plugin::TtRenderer::Helper->new(ctx => $c);

    # fixes for t/lite_app_with_default_layouts.t
    unless ($c->stash->{layout}) {
        $c->stash->{content} ||= $c->stash->{'mojo.content'}->{content};
    }

    my @params = ({%{$c->stash}, c => $c, h => $helper}, $output, {binmode => ':utf8'});
    my $provider = $self->tt->{SERVICE}->{CONTEXT}->{LOAD_TEMPLATES}->[0];
    $provider->options($options);

    my $ok = do {
        if (defined $inline) {
            $self->tt->process(\$inline, @params);
        }
        else {
            my @ret = $provider->fetch($t);

            if (not defined $ret[1]) {
                $self->tt->process($ret[0], @params);
            }
            elsif (not defined $ret[0]) { # not found
                return;
            }
            else { # error
                return if $! == ENOENT && (not ref $ret[0]); # not found when not blessed exception
                die $ret[0];
            }
        }
    };

    # Error
    die $self->tt->error unless $ok;
}

1;    # End of Mojolicious::Plugin::TtRenderer::Engine

package
  Mojolicious::Plugin::TtRenderer::Helper;

use strict;
use warnings;

use base 'Mojo::Base';

our $AUTOLOAD;

__PACKAGE__->attr('ctx');

sub AUTOLOAD {
    my $self = shift;

    my $method = $AUTOLOAD;

    return if $method =~ /^[A-Z]+?$/;
    return if $method =~ /^_/;
    return if $method =~ /(?:\:*?)DESTROY$/;

    $method = (split /::/, $method)[-1];

    die qq/Unknown helper: $method/ unless $self->ctx->app->renderer->helpers->{$method};

    $self->ctx->$method(@_);
}

1;

package
  Mojolicious::Plugin::TtRenderer::Provider;

use strict;
use warnings;

use base 'Template::Provider';
use Scalar::Util 'weaken';

sub new {
    my $class = shift;
    my %params = @_;

    my $renderer = delete $params{renderer};

    my $self = $class->SUPER::new(%params);
    $self->renderer($renderer);
    $self;
}

sub renderer      { @_ > 1 ? weaken($_[0]->{renderer} = $_[1]) : $_[0]->{renderer} }
sub options       { @_ > 1 ? $_[0]->{options}       = $_[1] : $_[0]->{options} }

sub _template_modified {
    my($self, $template) = @_;
    $self->SUPER::_template_modified($template) || $template =~ /^templates(?:\/|\\)/ || undef;
}

sub _template_content {
    my $self = shift;
    my ($path) = @_;

    my $options = delete $self->{options};

    # Convert backslashes to forward slashes to make inline templates work on Windows
    $path =~ s/\\/\//g;
    my ($t) = ($path =~ m{templates\/(.*)$});

    if (-r $path) {
        return $self->SUPER::_template_content(@_);
    }

    my $data;
    my $error = '';

    # Try DATA section
    if(defined $options) {
        $data = $self->renderer->get_data_template($options);
    } else {
        foreach my $class (@{ $self->renderer->classes }) {
            $data = Mojo::Loader::data_section($class, $t);
            last if $data;
        }
    }

    unless($data) {
        $data = '';
        $error = "$path: not found";
    }
    wantarray ? ($data, $error, time) : $data;  ## no critic (Freenode::Wantarray)
}

1;

__END__

=encoding utf-8

=begin stopwords

Bjørn
Szász
Árpád
Романов
Сергей

=end stopwords

=head1 SYNOPSIS

Add the handler:

 sub startup {
     ...
 
     # Via mojolicious plugin
     $self->plugin(tt_renderer => {template_options => {FILTERS => [ ... ]}});
 
     # Or manually
     use Mojolicious::Plugin::TtRenderer::Engine;
 
     my $tt = Mojolicious::Plugin::TtRenderer::Engine->build(
         mojo => $self,
         template_options => {
             PROCESS  => 'tpl/wrapper',
             FILTERS  => [ ... ],
             UNICODE  => 1,
             ENCODING => 'UTF-8',
         }
     );
 
     $self->renderer->add_handler( tt => $tt );
 }

Template parameter are taken from C<$c-E<gt>stash>.

=head1 DESCRIPTION

See L<Mojolicious::Plugin::TtRenderer> for details on the plugin interface to this module.

This module provides an engine for the rendering of L<Template Toolkit|Template> templates
within a Mojolicious context.  Templates may be, stored on the local file system, provided
inline by the controller or included in the C<__DATA__> section.  Where possible this modules
attempts to provide a TT analogue interface to the L<Perlish templates|Mojo::Template> which
come with Mojolicious.

=head1 RENDERING

The template file for C<"example/welcome"> would be C<"templates/welcome.html.tt">.

When template file is not available rendering from C<__DATA__> is attempted.

 __DATA__
 
 @@ welcome.html.tt
 Welcome, [% user.name %]!

Inline template is also supported:

 $self->render(inline => '[% 1 + 1 %]', handler => 'tt');

=head1 HELPERS

Helpers are exported automatically under C<h> namespace.

 [% h.url_for('index') %]

=head1 METHODS

=head2 build

This method returns a handler for the Mojolicious renderer.

Supported parameters are

=over 4

=item mojo
C<build> currently uses a C<mojo> parameter pointing to the base class (Mojo).
object. When used the INCLUDE_PATH will be set to

=item template_options

A hash reference of options that are passed to Template->new().  Note that if you
specify an C<INCLUDE_PATH> through this option it will remove the DATA section
templates from your path.  A better way to specify an C<INCLUDE_PATH> if you also
want to use DATA section templates it by manipulating the L<Mojolicious::Renderer>
path.

=item cache_dir

Absolute or relative dir to your app home, to cache processed versions of your
templates. Will default to a temp-dir if not set.

=back

=head1 SEE ALSO

L<Mojolicious::Plugin::TtRenderer>,
L<Mojolicious>,
L<Mojolicious::Guides>,
L<http://mojolicious.org>.

=cut
